File: scripts/set_frontend_api_base.py
#!/usr/bin/env python3
"""Simple helper to inject the runtime API base into the static frontend."""

import argparse
import os
import re
from pathlib import Path


def update_api_base(index_path: Path, api_base: str) -> bool:
    pattern = re.compile(r'(<meta\s+name="api-base"\s+content=")([^"]*)(")', re.IGNORECASE)
    html = index_path.read_text(encoding="utf-8")
    updated_html, count = pattern.subn(rf'\1{api_base}\3', html, count=1)
    if count == 0:
        return False
    index_path.write_text(updated_html, encoding="utf-8")
    return True


def main():
    parser = argparse.ArgumentParser(description="Inject API base URL into the static frontend.")
    parser.add_argument("--api-base", dest="api_base", default=os.environ.get("API_BASE_URL"), help="API base URL (e.g. https://service.onrender.com/api)")
    parser.add_argument("--index", dest="index_path", default="frontend/index.html", help="Path to the frontend index file.")
    args = parser.parse_args()

    if not args.api_base:
        raise SystemExit("API base URL not provided. Pass --api-base or set API_BASE_URL.")

    index_path = Path(args.index_path)
    if not index_path.exists():
        raise SystemExit(f"Cannot find index file at {index_path}")

    if not update_api_base(index_path, args.api_base.rstrip("/")):
        raise SystemExit("Failed to locate <meta name=\"api-base\"> tag in the index file.")

    print(f"Updated API base to {args.api_base.rstrip('/')}")


if __name__ == "__main__":
    main()

File: scripts/set_frontend_api_base.js
#!/usr/bin/env node
"use strict";

const fs = require("fs");
const path = require("path");

function readArgValue(args, name) {
  const exactIndex = args.indexOf(name);
  if (exactIndex !== -1) {
    if (exactIndex === args.length - 1) {
      throw new Error(`Missing value for ${name}`);
    }
    return args[exactIndex + 1];
  }

  const prefix = `${name}=`;
  const prefixed = args.find((arg) => arg.startsWith(prefix));
  if (prefixed) {
    return prefixed.slice(prefix.length);
  }

  return undefined;
}

function updateApiBase(htmlPath, apiBase) {
  const metaRegex = /(<meta\s+name="api-base"\s+content=")([^"]*)(")/i;
  const html = fs.readFileSync(htmlPath, "utf8");
  const match = html.match(metaRegex);
  if (!match) {
    return false;
  }
  const updatedHtml = html.replace(metaRegex, (_match, before, _content, after) => `${before}${apiBase}${after}`);
  fs.writeFileSync(htmlPath, updatedHtml, "utf8");
  return true;
}

function main() {
  const args = process.argv.slice(2);
  const apiBaseArg = readArgValue(args, "--api-base");
  const indexArg = readArgValue(args, "--index");
  const frontendDirArg = readArgValue(args, "--frontend-dir") ?? "frontend";

  const rawApiBase = apiBaseArg ?? process.env.API_BASE_URL;
  if (!rawApiBase) {
    throw new Error("API base URL not provided. Pass --api-base or set API_BASE_URL.");
  }

  const apiBase = rawApiBase.replace(/\/+$/, "");

  let htmlPaths = [];
  if (indexArg) {
    const resolved = path.resolve(indexArg);
    if (!fs.existsSync(resolved)) {
      throw new Error(`Cannot find index file at ${resolved}`);
    }
    htmlPaths = [resolved];
  } else {
    const frontendDir = path.resolve(frontendDirArg);
    if (!fs.existsSync(frontendDir) || !fs.statSync(frontendDir).isDirectory()) {
      throw new Error(`Cannot find frontend directory at ${frontendDir}`);
    }
    htmlPaths = fs
      .readdirSync(frontendDir)
      .filter((name) => name.toLowerCase().endsWith(".html"))
      .sort()
      .map((name) => path.join(frontendDir, name));

    if (htmlPaths.length === 0) {
      throw new Error(`No HTML files found under ${frontendDir}`);
    }
  }

  let updated = 0;
  for (const htmlPath of htmlPaths) {
    if (updateApiBase(htmlPath, apiBase)) {
      updated += 1;
    }
  }

  if (updated === 0) {
    throw new Error('Failed to locate <meta name="api-base"> tag in the provided HTML file(s).');
  }

  if (htmlPaths.length === 1) {
    process.stdout.write(`Updated API base to ${apiBase}\n`);
  } else {
    process.stdout.write(`Updated API base to ${apiBase} in ${updated} file(s)\n`);
  }
}

try {
  main();
} catch (error) {
  process.stderr.write(`${error.message ?? String(error)}\n`);
  process.exit(1);
}

